bst.py

```diff
@@ -131,17 +140,152 @@ def get_dot(self):
             )
         ))
 
+    def in_order(self):
+        return self._in_order(self.root)
+
+    def _in_order(self, leaf):
+        if leaf is None:
+            return
+        for value in self._in_order(leaf.left):
+            yield value
+        yield leaf.value
+        for value in self._in_order(leaf.right):
+            yield value
+
+    def pre_order(self):
+        return self._pre_order(self.root)
+
+    def _pre_order(self, leaf):
+        if leaf is None:
+            return
+        yield leaf.value
+        for value in self._pre_order(leaf.left):
+            yield value
+        for value in self._pre_order(leaf.right):
+            yield value
+
+    def post_order(self):
+        return self._post_order(self.root)
+
+    def _post_order(self, leaf):
+        if leaf is None:
+            return
+        for value in self._post_order(leaf.left):
+            yield value
+        for value in self._post_order(leaf.right):
+            yield value
+        yield leaf.value
+
+    def breadth_traversal(self):
+        x = deque()
+        x.append(self.root)
+        while x:
+            leaf = x.popleft()
+            yield leaf.value
+            if leaf.left:
+                x.append(leaf.left)
+            if leaf.right:
+                x.append(leaf.right)
+
+    def delete(self, val):
+        self.root = self._delete(val, self.root)
+        return None
+
+    def _delete(self, val, leaf):
+        def _descendants(leaf):
+            if leaf.left:
+                return _descendants(leaf.left)
+            else:
+                return leaf.value
+
+        if not leaf:
+            return None
+
+        if leaf.value == val:
+            self._size -= 1
+            if leaf.left and leaf.right:
+                leaf.value = _descendants(leaf.right)
+                leaf.right = self._delete(leaf.value, leaf.right)
+                return leaf
+            elif leaf.left and not leaf.right:
+                return leaf.left
+            elif not leaf.left and leaf.right:
+                return leaf.right
+            else:
+                return None
+
+        elif leaf.value < val:
+            if leaf.right:
+                leaf.right = self._delete(val, leaf.right)
+            return leaf
+
+        else:
+            if leaf.left:
+                leaf.left = self._delete(val, leaf.left)
+            return leaf
+
+    def l_rotate(self, node):
+        x = node.left
+        node.left = x.right
+        x.right = node
+        node.height = max(self.height(node.right), self.height(node.left)) + 1
```

Comment:
as discussed in class, making this a method on the node class would reduce repetition.